edb/edgeql/compiler/inference/cardinality.py

```diff
@@ -1035,15 +1035,24 @@ def __infer_insert_stmt(
         ir.result, is_mutation=True, scope_tree=new_scope, ctx=ctx
     )
 
-    if ir.on_conflict:
-        for part in [ir.on_conflict.select_ir, ir.on_conflict.else_ir]:
-            if part:
-                infer_cardinality(part, scope_tree=scope_tree, ctx=ctx)
-
     assert not ir.iterator_stmt, "InsertStmt shouldn't ever have an iterator"
 
     # INSERT without a FOR is always a singleton.
-    return ONE
+    if not ir.on_conflict:
+        return ONE
+    # ... except if UNLESS CONFLICT is used
+    else:
+        if ir.on_conflict.select_ir:
+            infer_cardinality(
```

Comment:
Should it be `card = infer_cardinality(...`?

Reply:
Aha! I had thought that we didn't need to worry about the cardinality of the select_ir because it would always be AT_MOST_ONE, but now that we support select_irs that look at multiple different exclusive constraints, it can be MANY!

`card = infer_cardinality(...` isn't quite what we need to do, since it shouldn't have any effect at all if there is no `ELSE`, but we do need to take it into account

Good catch!